apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -48,19 +49,59 @@ ${context}
  *
  * @param client The OpenAI client to use.
  * @param prompt The user's question generation prompt.
+ * @param promptUserInput The user's indication of how to create student input boxes.
+ * @param mandatoryElementIds Elements that we must pull documentation for.
  * @param authnUserId The user's authenticated user ID.
  * @returns A string of all relevant context documents.
  */
-async function makeContext(client: OpenAI, prompt: string, authnUserId: string): Promise<string> {
+async function makeContext(
+  client: OpenAI,
+  prompt: string,
+  promptUserInput: string | undefined,
+  mandatoryElementIds: string[],
+  authnUserId: string,
+): Promise<string> {
   const embedding = await createEmbedding(client, prompt, openAiUserFromAuthn(authnUserId));
+  const mandatoryElements =
+    mandatoryElementIds.length > 0
+      ? await queryRows(
+          sql.select_documents_by_chunk_id,
+          {
+            doc_path: 'docs/elements.md',
+            chunk_ids: mandatoryElementIds,
+          },
+          QuestionGenerationContextEmbeddingSchema,
+        )
+      : [];
+
+  const numElements = Math.max(5 - mandatoryElements.length, 0);
 
   const docs = await queryRows(
     sql.select_nearby_documents,
-    { embedding: vectorToString(embedding), limit: 5 },
+    { embedding: vectorToString(embedding), limit: numElements },
     QuestionGenerationContextEmbeddingSchema,
   );
 
-  return docs.map((doc) => doc.doc_text).join('\n\n');
+  if (promptUserInput !== undefined) {
+    const elementDoc = await queryRow(
+      sql.select_nearby_documents_from_file,
+      {
+        embedding: vectorToString(embedding),
+        doc_path: 'docs/elements.md',
+        limit: 1,
+      },
+      QuestionGenerationContextEmbeddingSchema,
+    );
```

Comment:
Please add comments. I don't understand why we do this, and why we only do this if `promptUserInput` is truthy.

Reply:
Changed in 09ebd145316f730bdcbac5525887fda484b63483